### DISCUSSION TITLE
What's the recommended mixed project layout for multiple Rust extensions?

### DISCUSSION THREAD

**Discussion by DavidAntliff** [PyO3/maturin#2126] [Q&A] [answered]
Issue https://github.com/PyO3/maturin/issues/490 led to a change in the recommended mixed project layout. Specifically, for a Python subdir within a Rust top-dir:

```
my-project
├── Cargo.toml
├── python
│   └── my_project
│       ├── __init__.py
│       └── bar.py
├── pyproject.toml
├── README.md
└── src
    └── lib.rs
```

How is this intended to work when there are _multiple_ independent Rust extensions in a single Python project? I.e. multiple sub-directories with independent `Cargo.toml` files.

Is the following layout supported?

```
my-project
├── python
│   └── my_project
│       ├── __init__.py
│       └── bar.py
├── pyproject.toml
├── README.md
└── rust
    ├── my_project1
    |   ├── Cargo.toml
    |   └── src
    |       └── lib.rs
    └── my_project2
        ├── Cargo.toml
        └── src
            └── lib.rs
```

**Comment by messense** [comment, marked as answer]
No support for multiple extension modules in multiple Cargo projects yet, however you can use one Cargo project and add other modules as submodules.

For example: https://github.com/pyca/cryptography/blob/541beda62fde26ee16f98092e553a17cac8e8943/src/rust/src/backend/mod.rs#L28-L53

**Reply by DavidAntliff** [reply]
Thanks, that seems like a good workaround.

Can I ask one more thing - I see in the [User Guide](https://www.maturin.rs/project_layout#alternate-python-source-directory-src-layout) there's an alternative layout where the (single) Rust project is moved into a subdirectory, and:

> maturin will detect the following src layout **automatically**

(My emphasis)

Is the magic the fact the subdir is specifically called "rust", or will `maturin` go looking for a subdir containing a `Cargo.toml`? If so, how far does it look? Can this be configured?

```
my-rust-and-python-project
├── src  # put python code in src folder
│   └── my_project
│       ├── __init__.py
│       └── bar.py
├── pyproject.toml
├── README.md
└── rust # put rust code in rust folder
    |── Cargo.toml
    └── src
        └── lib.rs
```

**Reply by messense** [reply]
It simply looks for `src/<project_name>` and `rust` subdirs, only python source dir can be configured ATM, see https://www.maturin.rs/config#maturin-options

**Reply by DavidAntliff** [reply]
Ok, useful info, thank you.
